supplementary/alica_ros1/alica_ros_turtlesim/libalica-turtlesim/src/FourCorners.cpp

```diff
@@ -0,0 +1,51 @@
+#include "FourCorners.h"
+#include "FourCornersSummand.h"
+#include "engine/USummand.h"
+#include "engine/blackboard/Blackboard.h"
+#include "ros/ros.h"
+namespace turtlesim
+{
+
+std::shared_ptr<FourCornersUtilityFunction> FourCornersUtilityFunction::create(alica::UtilityFunctionContext& context)
+{
+    return std::make_shared<FourCornersUtilityFunction>(context);
+}
+
+/*
+    The utility function combines three components to determine the best assignment:
+    1. Priority (prio): A measure of how important the assignment is considering agent role and it's priority to take up a task.
+    2. Similarity (sim): A measure of how close the new assignment is to the previous assignment.
+    3. Custom component (summand): A user-defined component to add extra criteria.
+
+    Utility = priorityWeight * prio(assignment) + similarityWeight * sim(assignment, oldAssignment) + customWeight * summand(assignment)
+
+    The utility threshold is set to 0.750 for this plan (see FourCorner.pml for details).
+
+    If summand(assignment) returns 0, it indicates that the current assignment is not ideal. In this case, it can only be accepted if it is highly similar to
+   the oldAssignment. When the current assignment is an improvement but deviates considerably from the oldAssignment, the utility score will not surpass the
+   threshold, leading to the assignment's rejection.
+
+    Initially, only suitable assignments are selected. In later iterations, the algorithm searches for new assignments that are both better and similar to the
+   selected ones.
+
+    The similarityWeight helps the solution converge faster by not being too greedy in accepting better solutions. It also helps maintain stability among agents
```

Comment:
I'd say it makes conversion slower, precisely because it's more stable. Can you elaborate?

Reply:
I think it is making convergence faster, by reducing oscillating behavior between assignments. Say for example we are executing an assignment and while performing the assigned tasks a better assignment is found, if we keep on accepting the better assignment we might encounter some situation where the system just oscillates between assignments as there can be multiple best assignments, and can delay to converge on a single assignment. So the point is to reduce this tendency and stick to what assignment we are already performing and only switch if similar assignments are found which will converge faster as basically system would be exploiting more.

Reply:
Ah, because the greedy approach oscillates too much :+1: